### ISSUE TITLE
Newest version of lua-resty-string causes aes encoding to fail

### ISSUE THREAD

**Issue by peturorri** [bungle/lua-resty-session#40]
With the newest version of lua-resty-string the following error is given by the resty.aes module:

    error: salt must be 8 characters or nil

causing the sample code to fail (aes object is nil).

See this commit:
https://github.com/openresty/lua-resty-string/commit/82ff0e3660305078f57a5986be27c530a2141def

**Comment by bungle** [comment]
@peturorri thanks for letting me know. I will investigate on it.

**Comment by bungle** [comment]
@peturorri, it is now fixed and released. If you happen to have any more problems with it, please don't hesitate to reopen this issue.

**Comment by peturorri** [comment]
Looks good. Thanks for the quick response.

**Comment by gene1wood** [comment]
This bug manifested for me with the error

```
runtime error: /usr/share/lua/5.1/resty/session/ciphers/aes.lua:46: attempt to index a nil value
```

because the call to [`aes:new`](https://github.com/bungle/lua-resty-session/blob/d006e524feb4d53880461c8c97bf8c7560a4b6c7/lib/resty/session/ciphers/aes.lua#L63) was [returning `nil`](https://github.com/openresty/lua-resty-string/blob/a55eb9e3e0f08e1797cd5b31ccea9d9b05e5890b/lib/resty/aes.lua#L163)

Just adding this in case it helps anyone else find there way to this bug report.

**Comment by bungle** [comment]
@gene1wood, thank you for this, that could be improved.

**Comment by gene1wood** [comment]
@bungle Ya, maybe test for nil, and capture the second argument returned indicating the reason for the failure
* [bad key length](https://github.com/openresty/lua-resty-string/blob/a55eb9e3e0f08e1797cd5b31ccea9d9b05e5890b/lib/resty/aes.lua#L147)
* [bad iv](https://github.com/openresty/lua-resty-string/blob/a55eb9e3e0f08e1797cd5b31ccea9d9b05e5890b/lib/resty/aes.lua#L140)

etc

**Comment by bungle** [comment]
@gene1wood, yes, I will go through all the code and try to improve error reporting to make this not runtime error. Did you change the defaults when you got this error?

**Comment by bungle** [comment]
I reopened this to remind me that this could be still made better.

**Comment by gene1wood** [comment]
@bungle I'm not sure I understand what you mean by "Did you change the defaults when you got this error?", can you elaborate?

**Comment by bungle** [comment]
@gene1wood, I mean why did you get:
```
runtime error: /usr/share/lua/5.1/resty/session/ciphers/aes.lua:46: attempt to index a nil value
```

Default cipher should work, so I was just thinking, how did you end up with that? E.g. did you try different cipher or was the error because of something else (e.g. using older `lua-resty-session` or non-`OpenResty` build of Nginx or linking to a certaine `OpenSSL`?

**Comment by gene1wood** [comment]
Got it.

I was using `lua-resty-session` `2.8` with `open-resty` `1.13.6.1` because the OS package for `open-resty` was updated but the luarocks package for `lua-resty-session` was not. The fix was to update `lua-resty-session` to `2.19` with luarocks.
